source/adios2/toolkit/transport/file/FileFStream.cpp

```diff
@@ -23,19 +23,50 @@ FileFStream::FileFStream(helper::Comm const &comm, const bool debugMode)
 {
 }
 
-void FileFStream::Open(const std::string &name, const Mode openMode)
+void FileFStream::WaitForOpen()
 {
+    if (m_IsOpening)
+    {
+        if (m_OpenFuture.valid())
+        {
+            m_OpenFuture.get();
+        }
+        m_IsOpening = false;
+        CheckFile(
+            "couldn't open file " + m_Name +
+            ", check permissions or path existence, in call to POSIX open");
+        m_IsOpen = true;
+    }
+}
+
+void FileFStream::Open(const std::string &name, const Mode openMode,
+                       const bool async)
+{
+    auto lf_AsyncOpen = [&](const std::string &name) -> void {
+        ProfilerStart("open");
+        m_FileStream.open(name, std::fstream::out | std::fstream::binary |
+                                    std::fstream::trunc);
+        ProfilerStop("open");
+    };
```

Comment:
Should this be setup to work with both writing and reading?
```c++
    auto lf_AsyncOpenWrite = [&](const std::string &name) -> void {
        ProfilerStart("open");
        m_FileStream.open(name, std::fstream::out | std::fstream::binary |
                                    std::fstream::trunc);
        ProfilerStop("open");
    };
    auto lf_AsyncOpenRead = [&](const std::string &name) -> void {
        ProfilerStart("open");
        m_FileStream.open(name, std::fstream::in | std::fstream::binary);
        ProfilerStop("open");
    };
```

Reply:
It certainly could be.  However, in the past the use case has been for writers, so async open was only implemented for them.  The use case was that some codes would open output files at init time, but then do a timesteps worth of computation before it needed to write them, so if Open() was going to take a while, doing it async pays off by hiding that delay behind computation.  On the reader side, the analysis codes are maybe less likely to have something to do between opening the file and wanting to get data to work on.  However, unlike the prior versions, pushing this down results in an async task per OS file, rather than one per ADIOS file, so there may be some use in doing open that have substreams in parallel...   I'd defer to Norbert to opine on why this wasn't implemented for readers in the past.

Reply:
It was simply not required because we don't know any case where the reader would pre-open a file. But it is certainly not a bad idea to support it if it is not expensive for ALL reading

Reply:
It wouldn't be expensive in any way, but there is another consequence.  It does move knowledge of whether or not an open() succeeded or failed *out* of the ADIOS-level Open() and into the next operation on that file.  I.E. Open with async will always report success, but the next write(), read(), etc. will raise any exception that the Open might have.  For create-mode writers I think that's probably acceptable as a default (and matches the semantics from before the TSAN mods).  For readers, we probably have to be more careful.  If system-level open() is used as an existence test, then doing it async would bollox that up.

Reply:
Actually, I was wrong here.  If we tried to delay reader-side system-level open(), we wouldn't get to to the point where we'd not know if a file existed until after the high-level Open(), because the first thing the BP reader engines do after open is to get the file size and read the metadata.  So if the system-level open() was async, we'd really be waiting on it a few lines later and it would all be relatively pointless.  To get truely async opens, we'd have to restructure the readers a fair bit so that we read the metadata later, maybe on demand at first use.  Not impossible, but likely not worth the trouble without a compelling use case.  For the writers though, after open we're building metadata data structures in memory for the rest of the high-level Open() call and actual file writes or accesses don't occur until later high-level operations happen.    Given this, my suggestion would be to run with this version that makes only the writes async.  This restores prior functionality (async system-level writer open()) in a way that makes TSAN happy.  If we develop a use-case for async high-level reader open(), we can tackle that in a different PR (that would necessarily involve significant changes in BPxReader code as well).  @pnorbert and @chuckatkins, if you concur maybe we can resolve this conversation and merge?

Reply:
Sounds good to me.  It makes sense to keep this in the transport anyways.  The only change is request is to rename `lf_AsyncOpen` to `lf_AsyncOpenWrite` just so it's clear that this implementation isn't for both modes.

Reply:
I'll rename.  But while I was doing further due diligence, I discovered that while this patch restores the async Open() functionality for the BP3 writer, it does not do so for BP4, because BP4Writer does not completely avoid Write() in its InitTransports() sequence (change related to the new index files in BP4).  My temptation would be to still merge this (after naming change) as it fixes BP3, open an issue on the BP4 async open functionality, and maybe open a 2nd issue on how to test this in CI. If it's important enough to implement, we should test. I've been manually testing by temporarily adding a multi-second delay into the AsyncOpen task, printing "low-level Open done" when it completes and adding a "high-level Open done" when the transport Open completes.  It's working if the "high-level Open" completes before the low-level open.  But I don't commit those delays and output lines obviously.  This is functionality that might be easily broken by changes in the BP engines, can we actually test it in CI?

Reply:
>  My temptation would be to still merge this (after naming change) as it fixes BP3, open an issue on the BP4 async open functionality

Agreed

> maybe open a 2nd issue on how to test this in CI.
> ...
> This is functionality that might be easily broken by changes in the BP engines, can we actually test it in CI?

I can't really think of a reliable way.  Any tests that are timing based are inherently unstable on CI given the reduced adn low priority environments they run in.

Reply:
lgtm

Reply:
Messed up those mods.  Trying again shortly.

Reply:
This is merged now, but I have to say that I find it ironic that we can't test functionality designed to help in environments where OS functions might be slow because our CI environment might be slow...